Replace Write-Host with custom function - Attempt #2

FabienTschanz opened the pull request:
#### Pull Request (PR) description
This PR replaces calls to `Write-Host` with a new call to `Write-M365DSCHost`, which outputs the message to the console for interactive users and to the verbose stream when running in a non-interactive shell, e.g. during Azure Automation.

#### This Pull Request (PR) fixes the following issues
None.

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [ ] Resource parameter descriptions added/updated in the schema.mof.
- [ ] Resource documentation added/updated in README.md.
- [ ] Resource settings.json file contains all required permissions.
- [ ] Examples appropriately added/updated.
- [ ] Unit tests added/updated.
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

FabienTschanz committed:
Fix commit write with empty stashed messages

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

FabienTschanz commented:
@NikCharlebois The issue was that a null-index access occurred because of a missing check in M365DSCUtil.psm1: 

```powershell
if ([Environment]::UserInteractive)
    {
        if ($CommitWrite -and $Script:M365DSCHostMessages.Count -gt 0) # Condition -and was missing
        {
            for ($i = 0; $i -lt $Script:M365DSCHostMessages.Count - 1; $i++)
            {
                Write-Host -Object $Script:M365DSCHostMessages[$i].Message -ForegroundColor $Script:M365DSCHostMessages[$i].ForegroundColor -NoNewline
            }
            Write-Host -Object $Script:M365DSCHostMessages[-1].Message -ForegroundColor $Script:M365DSCHostMessages[-1].ForegroundColor -NoNewline
            $Script:M365DSCHostMessages = @()
        }
        Write-Host -Object $Message -ForegroundColor $ForegroundColor
    }
```

Just ran an export again and it works as intended.

FabienTschanz commented:
I understand that this is a massive PR. Therefore, it shouldn't be merged one day before a release but rather just after a release. This way, we have up to one week to check for issues and don't have to revert the PR 5 minutes after merging. Getting all changes ready again was a real pitb to be blunt, for a change that affected only one change if it had been kept in the Dev branch.

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

FabienTschanz committed:
Update new resources to align with host output

FabienTschanz commented:
@ykuijs @NikCharlebois Is it possible to merge this one and check until next Wednesday, if there are issues we can fix them and if it still doesn't work, we'll revert it. Does that work for you?

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

NikCharlebois commented:
I will park this one again for today's release as we haven't had a chance to fully review it and the potential impacts. Will try to merge on Friday.

FabienTschanz commented:
@NikCharlebois No problem, please take your time. If you like, you can also merge it for the next breaking release in April if that suits you and your team better.

Commit:
Merge branch 'Dev' into feat/ref-write-host

ykuijs commented:
GitHub doesn't allow me to resolve the conflicts myself. @FabienTschanz would be great if you have the time to update this PR with all changes from dev. I will merge it afterwards.

FabienTschanz committed:
Merge branch 'Dev' into feat/ref-write-host

FabienTschanz committed:
Update new resources for Write-Host changes

FabienTschanz commented:
@ykuijs I updated the PR with the latest changes and also changed the newly added resources to behave the same as all the others. Waiting on the unit tests to complete if there is an issue. Thank you 👍
